### ISSUE TITLE
Data loss with floating-point `Histogram`s

### ISSUE THREAD

**Issue by yurivish** [JuliaStats/StatsBase.jl#1009]
The function `fit(Histogram, v; nbins, closed)` silently drops observations at the extremes of the data. 

Here's a quick simulation illustrating the bug using 100 batches of measurements, each with 1 to 3 decimal digits. We take readings in batches with 2 to 5 readings per batch and observe that StatsBase `Histogram`s quietly drop observations; 11 of the 100 runs lose at least one reading.

```julia
julia> using StatsBase, Random

julia> Random.seed!(2026); # feel free to tweak this hyperparameter; including for reproducibility

julia> total_collected = total_counted = batches_with_data_loss = 0;

julia> for batch in 1:100
           n = rand(2:5)
           digs = rand(1:3)
           readings = Float32[round(rand(), digits=digs) for _ in 1:n]
           global total_collected += length(readings)

           h = fit(Histogram, readings, nbins=rand(1:12), closed=rand((:left, :right)))
           c = sum(h.weights)
           global total_counted += c
           global batches_with_data_loss += (c != length(readings))
       end

julia> total_collected # total readings collected
364

julia> total_counted # total readings countd
352

julia> readings_dropped = total_collected - total_counted # number of individual data readings that were silently dropped
12

julia> batches_with_data_loss # number of batches with data loss
11
```

## AI-Suggested fix

`binindex` decides membership by `searchsortedlast` / `searchsortedfirst` against the returned range, so the only check that can't drift is one made against that range. Build it, then re-run the endpoint fix-up against `first(r)` / `last(r)`:

```julia
    r = StepRangeLen(Base.TwicePrecision{Float64}((start, divisor)),
                     Base.TwicePrecision{Float64}((step, divisor)),
                     Int(len))
    # The loops above ran in `F` with plain division; the edges the caller
    # gets are evaluated by the TwicePrecision range, and the two can differ
    # by an ulp. Re-check against the range itself, which is what `binindex` uses.
    lo64, hi64 = Float64(lo), Float64(hi)
    if closed == :right
        while lo64 <= first(r); r = StepRangeLen(r.ref - r.step, r.step, length(r) + 1); end
        while last(r) < hi64;   r = StepRangeLen(r.ref, r.step, length(r) + 1); end
    else
        while lo64 < first(r);  r = StepRangeLen(r.ref - r.step, r.step, length(r) + 1); end
        while last(r) <= hi64;  r = StepRangeLen(r.ref, r.step, length(r) + 1); end
    end
    return r
```

This replaces the final `StepRangeLen(...)` expression in `histrange(lo::F, hi::F, n, closed)`. It is type-stable, passes the existing `test/hist.jl` (148 tests), and takes the drop rate to 0 in 40,000-fit sweeps for `Float32`, `Float16` and `Float64` on [0,1) and `Float32` on [0,10). The only visible change is that an affected range gains one bin at the end that was losing data, which is exactly what the existing fix-up loops already do in the cases they catch.

For `Float64` input the `F` arithmetic and the range's arithmetic are still different: the guard evaluates `(start + (len-1)*step)/divisor` with one Float64 division, while the range evaluates `ref + (len-1)*step` in `TwicePrecision`, and for `closed=:right` the latter can come out one ulp lower (`0.28:0.02:0.39999999999999997` for `[0.3, 0.4]`, `nbins=7`). The re-check above covers it with no special-casing.

Two narrower fixes were tried and are not enough:
- Promoting `lo`/`hi` to `Float64` at the top fixes `Float32`/`Float16` but leaves the `Float64` `closed=:right` drops untouched.
- Building the range in `TwicePrecision{F}` still drops `Float32` observations (the `F` guard and the `TwicePrecision{F}` range disagree the same way), leaves `Float64` unchanged, and throws a `BoundsError` for `Float16`.

Regression test: `sum(fit(Histogram, v; nbins=n, closed=c).weights) == length(v)` for every auto-edge call, over `Float16`/`Float32`/`Float64` and both `closed`.
